tests/integration_tests/datasources/test_lxd_hotplug.py

```diff
@@ -0,0 +1,121 @@
+import json
+
+import pytest
+
+from cloudinit import safeyaml
+from cloudinit.subp import subp
+from tests.integration_tests.clouds import ImageSpecification
+from tests.integration_tests.decorators import retry
+from tests.integration_tests.instances import IntegrationInstance
+
+USER_DATA = """\
+#cloud-config
+updates:
+  network:
+    when: ["hotplug"]
+"""
+
+UPDATED_NETWORK_CONFIG = """\
+version: 2
+ethernets:
+    eth0:
+        dhcp4: true
+    eth2:
+        dhcp4: true
+"""
+
+
+@retry()
+def ensure_hotplug_exited(client):
+    assert "cloud-init" not in client.execute("ps -A")
+
+
+def get_parent_network(instance_name: str):
+    lxd_network = json.loads(
+        subp("lxc network list --format json".split()).stdout
+    )
+    for net in lxd_network:
+        if net["type"] == "bridge" and net["managed"]:
+            if f"/1.0/instances/{instance_name}" in net.get("used_by", []):
+                return net["name"]
+    return "lxdbr0"
+
+
+@pytest.mark.lxd_container
+@pytest.mark.lxd_vm
+@pytest.mark.user_data(USER_DATA)
+class TestLxdHotplug:
+    @pytest.fixture(autouse=True, scope="class")
+    def class_teardown(self, class_client: IntegrationInstance):
+        # We need a teardown here because on IntegrationInstance teardown,
+        # if KEEP_INSTANCE=True, we grab the instance IP for logging, but
+        # we're currently running into
+        # https://github.com/canonical/pycloudlib/issues/220 .
+        # Once that issue is fixed, we can remove this teardown
+        yield
+        name = class_client.instance.name
+        subp(f"lxc config device remove {name} eth1".split())
+        subp(f"lxc config device remove {name} eth2".split())
+
+    def test_no_network_change_default(
+        self, class_client: IntegrationInstance
+    ):
+        client = class_client
+        assert "eth1" not in client.execute("ip address")
+        pre_netplan = client.read_from_file("/etc/netplan/50-cloud-init.yaml")
+        parent_network = get_parent_network(client.instance.name)
+        assert subp(
+            f"lxc config device add {client.instance.name} eth1 nic name=eth1 "
+            f"nictype=bridged parent={parent_network}".split()
+        )
+        ensure_hotplug_exited(client)
+        post_netplan = client.read_from_file("/etc/netplan/50-cloud-init.yaml")
+        assert pre_netplan == post_netplan
+        ip_info = json.loads(client.execute("ip --json address"))
+        eth1s = [i for i in ip_info if i["ifname"] == "eth1"]
+        assert len(eth1s) == 1
+        assert eth1s[0]["operstate"] == "DOWN"
+
+    def test_network_config_applied(self, class_client: IntegrationInstance):
```

Comment:
These tests will fail on bionic and focal because NoCloud seed files are still written to `/var/lib/cloud/nocloud-seed` for Bionic and Focal.
To avoid that we need to create a /etc/cloud/cloud.cfg,d dropin that'll force LXD detection before NoCloud. Something like the following diff should get that in place (also not this diff tries to avoid the duplicate nics from one host attached to same managed bridge error on LXD 5.7:
```diff
diff --git a/tests/integration_tests/datasources/test_lxd_hotplug.py b/tests/integration_tests/datasources/test_lxd_hotplug.py
index b026dcc8b..6ec8d96d0 100644
--- a/tests/integration_tests/datasources/test_lxd_hotplug.py
+++ b/tests/integration_tests/datasources/test_lxd_hotplug.py
@@ -41,6 +41,20 @@ def get_parent_network(instance_name: str):
     return "lxdbr0"
 
 
+def _prefer_lxd_datasource_over_nocloud(client: IntegrationInstance):
+    """For hotplug support we need LXD datasource detected instead of NoCloud
+
+    Bionic and Focal still deliver nocloud-net seed files so override it
+    with /etc/cloud/cloud.cfg.d/99-detect-lxd-first.cfg
+    """
+    client.write_to_file(
+        "/etc/cloud/cloud.cfg.d/99-detect-lxd-first.cfg",
+        "datasource_list: [LXD, NoCloud]\n",
+    )
+    client.execute("cloud-init clean --logs")
+    client.restart()
+
+
 @pytest.mark.lxd_container
 @pytest.mark.lxd_vm
 @pytest.mark.user_data(USER_DATA)
@@ -56,20 +70,30 @@ class TestLxdHotplug:
         name = class_client.instance.name
         subp(f"lxc config device remove {name} eth1".split())
         subp(f"lxc config device remove {name} eth2".split())
+        subp(f"lxc network delete ci-test-br-eth1".split())
+        subp(f"lxc network delete ci-test-br-eth2".split())
 
     def test_no_network_change_default(
         self, class_client: IntegrationInstance
     ):
         client = class_client
+        if ImageSpecification.from_os_image().release in [
+            "bionic",
+            "focal",
+        ]:  # pyright: ignore
+            _prefer_lxd_datasource_over_nocloud(client)
         assert "eth1" not in client.execute("ip address")
         pre_netplan = client.read_from_file("/etc/netplan/50-cloud-init.yaml")
-        parent_network = get_parent_network(client.instance.name)
-        assert subp(
-            f"lxc network create {client.instance.name}-net --type=bridge ipv4.address=10.10.41.1/24 ipv4.nat=true".split()
-        )
+
+        networks = subp("lxc network list".split())
+        if "ci-test-br-eth1" not in networks.stdout:
+            assert subp(
+                f"lxc network create ci-test-br-eth1 --type=bridge"
+                " ipv4.address=10.10.41.1/24 ipv4.nat=true".split()
+            )
         assert subp(
             f"lxc config device add {client.instance.name} eth1 nic name=eth1 "
-            f"nictype=bridged parent={parent_network}".split()
+            f"nictype=bridged parent=ci-test-br-eth1".split()
         )
         ensure_hotplug_exited(client)
         post_netplan = client.read_from_file("/etc/netplan/50-cloud-init.yaml")
@@ -81,16 +105,17 @@ class TestLxdHotplug:
 
     def test_network_config_applied(self, class_client: IntegrationInstance):
         client = class_client
-        assert "eth2" not in client.execute("ip address")
-        pre_netplan = client.read_from_file("/etc/netplan/50-cloud-init.yaml")
-        assert "eth2" not in pre_netplan
         if ImageSpecification.from_os_image().release in [
             "bionic",
             "focal",
         ]:  # pyright: ignore
+            _prefer_lxd_datasource_over_nocloud(client)
             top_key = "user"
         else:
             top_key = "cloud-init"
+        assert "eth2" not in client.execute("ip address")
+        pre_netplan = client.read_from_file("/etc/netplan/50-cloud-init.yaml")
+        assert "eth2" not in pre_netplan
         assert subp(
             [
                 "lxc",
@@ -104,10 +129,15 @@ class TestLxdHotplug:
             client.read_from_file("/etc/netplan/50-cloud-init.yaml")
             == pre_netplan
         )
-        parent_network = get_parent_network(client.instance.name)
+        networks = subp("lxc network list".split())
+        if "ci-test-br-eth2" not in networks.stdout:
+            assert subp(
+                f"lxc network create ci-test-br-eth2 --type=bridge"
+                " ipv4.address=10.10.42.1/24 ipv4.nat=true".split()
+            )
         assert subp(
             f"lxc config device add {client.instance.name} eth2 nic name=eth2 "
-            f"nictype=bridged parent={parent_network}".split()
+            f"nictype=bridged parent=ci-test-br-eth2".split()
         )
         ensure_hotplug_exited(client)
         post_netplan = safeyaml.load(
```

The above changeset gets a pass for me on LXD 5.7 *edge" stream and focal